hide GL icon in MacOS dock until a window is shown

aplavin opened the pull request:
# Description

Fixes #4278 – a small but annoying bug which makes MacOS dock to show an icon per Julia process with GLMakie loaded.

Adds `show_gl_icon_in_dock(::Bool)` and calls it so that:
- By default, after loading GLMakie, the icon is hidden
- After `display()` is called, the icon appears

This should work intuitively for:
- scripts that export images/videos: no display, no icon
- notebooks that use GLMakie: no display, no icon
- scripts/repl that calls display(): with icon

The `show_gl_icon_in_dock` definitely works, I've used it in MakieExtra already. Not 100% sure if I put the calls in the right places though, could you please check? The GLMakie sequence of loading/activation/display is a bit convoluted for someone unfamiliar with it.

## Type of change

Delete options that do not apply:

- [ ] Bug fix (non-breaking change which fixes an issue)

## Checklist

- [x] Added an entry in CHANGELOG.md
- [ ] Added or changed relevant sections in the documentation
- [ ] Added unit tests for new algorithms, conversion methods, etc.
- [ ] Added reference image tests for new plotting functions, recipes, visual options, etc.

jkrumbiegel commented:
> For me the main priority is to resolve the practical issue of many useless icons popping up 

Yeah that's why I was suggesting the icon thing, because the use of an icon that looks like a terminal and says julia is also limited for users that don't expect this is how GLMakie appears.

I'm also not sure which is the correct point in the screen lifetime to do this, that's @SimonDanisch's area of expertise. I guess at every screen open or close it should be checked if any other screens are currently open, and the icon shown/hidden accordingly.

aplavin committed:
hide GL icon in MacOS dock

Fixes https://github.com/MakieOrg/Makie.jl/issues/4278

aplavin committed:
display icon when windows shown

aplavin committed:
dock shows makie icon

aplavin committed:
add changelog entry for macOS dock icon

aplavin committed:
fix Runic formatting

jkrumbiegel commented on GLMakie/src/GLMakie.jl:
```diff
@@ -93,12 +93,56 @@ include("gl_backend.jl")
 load_all_shaders(SHADER_DIR)
 WARN_ON_LOAD[] = true
 
+const MAKIE_ICON_PATH = normpath(joinpath(@__DIR__, "..", "assets", "icon.png"))
```
this is not relocatable I think

aplavin replied:
Not sure I understand – do you mean it breaks if the icon gets moved to another location? I assumed the `assets` folder is the natural place to put it.

asinghvi17 replied:
It will break if you use PackageCompiler to generate a system image with Makie, specifically because the absolute path gets baked in.  I think we use RelocatableFolders to get around this?

ffreyer replied:
Is `assetpath("icon.png")` safe to use?

asinghvi17 replied:
Yes, but only because it uses artifacts 😅 

https://github.com/MakieOrg/Makie.jl/blob/74e6f39ef9ff03859048d4a8167d2f7d58242908/Makie/src/Makie.jl#L89

jkrumbiegel committed:
fix(GLMakie): make dock icon path relocatable via MakieAssets artifact

The icon was loaded from GLMakie/assets/icon.png with @__DIR__, which
bakes the build-time path into the precompile cache and breaks
relocatability. Switch to Makie.assetpath("icon_transparent.png"),
which resolves through the MakieAssets artifact at runtime. The file
in the artifact is byte-identical to the one previously shipped under
GLMakie/assets/, so behavior is unchanged.

jkrumbiegel committed:
Merge branch 'master' into patch-10

jkrumbiegel committed:
fix(CHANGELOG): resolve leftover merge conflict markers, move dock icon entry to Unreleased

The previous master merge left literal conflict markers in CHANGELOG.md.
The dock icon entry belongs to the unreleased section (this PR is not in
0.24.10), while #5375 was released in 0.24.10 and stays there. Also
relinked the entry to PR #5223 instead of issue #4278.
